bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceUse.java

```diff
@@ -171,8 +210,147 @@ void decrementUse() {
 	/**
 	 * Reset the use count to zero.
 	 */
-	/* @GuardedBy("this") */
+	/* @GuardedBy("getLock()") */
 	void resetUse() {
 		useCount = 0;
 	}
+
+	/**
+	 * The ServiceUseLock for managing access to this ServiceUse.
+	 * <p>
+	 * Use {@link #lock()} to lock the ServiceUseLock in a try-with-resources
+	 * statement.
+	 * 
+	 * @return The ServiceUseLock for this ServiceUse.
+	 */
+	ServiceUseLock getLock() {
+		return lock;
+	}
+
+	/**
+	 * Acquires the ServiceUseLock of this ServiceUse.
+	 * 
+	 * If this ServiceUse is locked by another thread then the current thread lies
+	 * dormant until the lock has been acquired.
+	 * 
+	 * @return The unlocker, which is {@link AutoCloseable}, to unlock the
+	 *         ServiceUseLock of this ServiceUse.
+	 * @throws ServiceException If a deadlock with another ServiceUseLock is
+	 *                          detected.
+	 */
+	ServiceUseLock lock() {
+		Thread awaitingThread = null;
+		boolean interrupted = false;
+		try {
+			final ServiceUseLock useLock = getLock(); // local var to avoid multiple getfields
+			while (true) {
+				try {
+					if (useLock.tryLock(100_000_000L, TimeUnit.NANOSECONDS)) { // 100ms (but prevent conversion)
```

Comment:
I think the current code is correct. It takes the optimistic approach that the thread is not interrupted (which is the most common case). If the thread is interrupted, then the tryLock will throw an InterruptedException which will be caught and we record the interrupt status for setting upon return. The throwing of the exception clears the interrupt status and we loop back up to call tryLock again.

Reply:
Yes that was the intention.
Furthermore it is the only reliable approach because if we are unlucky the thread gets interrupted just after we called `Thread.interrupted()` and before `tryLock()` is called. So we have to be prepared anyway.